### PR TITLE
topdown/net: require prefix length for IPv6 in net.cidr_merge

### LINKED ISSUE #4596: Errors with net.cidr_merge
## Short description 
In short, the net.cidr_merge seems not produce the smallest possible set of CIDRs.

Examples:
I have IP addresses: ["192.168.1.254", "192.168.1.254"]
output: `192.168.1.0/24`
expect: `192.168.1.254/32`
Seems like it is actually expanding the IP range.

* OPA version
latest

## Additional context
In addition, the net.cidr_merge does not handle ipv6 very well. I have an ipv6 address that cannot be converted to ipv4(not mapped)
and 
```
2022/04/19 16:53:00 http: panic serving [::1]:35204: runtime error: index out of range [15] with length 0
goroutine 30 [running]:
net/http.(*conn).serve.func1()
        /usr/lib/go-1.18/src/net/http/server.go:1825 +0xbf
panic({0xe90d20, 0xc0003123d8})
        /usr/lib/go-1.18/src/runtime/panic.go:844 +0x258
github.com/open-policy-agent/opa/internal/cidr/merge.GetAddressRange({{0x170c3c0, 0x0, 0x0}, {0xc00025aeb0, 0x10, 0x10}})
        /home/alex_hu/go/pkg/mod/github.com/open-policy-agent/opa@v0.38.1/internal/cidr/merge/merge.go:117 +0x3c5
github.com/open-policy-agent/opa/internal/cidr/merge.partitionCIDR({{0xc00025ae30, 0x10, 0x10}, {0xc00025ae40, 0x10, 0x10}}, {{0x170c3c0, 0x0, 0x0}, {0xc00025aeb0, ...}})
        /home/alex_hu/go/pkg/mod/github.com/open-policy-agent/opa@v0.38.1/internal/cidr/merge/merge.go:225 +0x1c5
github.com/open-policy-agent/opa/internal/cidr/merge.RangeToCIDRs({0x170c3c0, 0x0, 0x0}, {0x170c3c0, 0x0, 0x0})
        /home/alex_hu/go/pkg/mod/github.com/open-policy-agent/opa@v0.38.1/internal/cidr/merge/merge.go:87 +0x6ba
github.com/open-policy-agent/opa/topdown.evalNetCIDRMerge({0xc0005cb640, 0x4, 0xc00001a060?})
        /home/alex_hu/go/pkg/mod/github.com/open-policy-agent/opa@v0.38.1/topdown/cidr.go:327 +0x39c
github.com/open-policy-agent/opa/topdown.builtinNetCIDRMerge({{0x107a4e0, 0xc00051aa20}, {0x107d928, 0xc0005cb420}, {0x10709e0, 0xc00011a1e0}, 0xc0005afb48, {0x10757a8, 0xc00025ac38}, 0xc000599da0, ...}, ...)
        /home/alex_hu/go/pkg/mod/github.com/open-policy-agent/opa@v0.38.1/topdown/cidr.go:288 +0x18b
github.com/open-policy-agent/opa/topdown.builtinErrorWrapper.func1({{0x107a4e0, 0xc00051aa20}, {0x107d928, 0xc0005cb420}, {0x10709e0, 0xc00011a1e0}, 0xc0005afb48, {0x10757a8, 0xc00025ac38}, 0xc000599da0, ...}, ...)
        /home/alex_hu/go/pkg/mod/github.com/open-policy-agent/opa@v0.38.1/topdown/builtins.go:126 +0x79
github.com/open-policy-agent/opa/topdown.evalBuiltin.eval({0xc000001c20, 0x16b6060, {{0x107a4e0, 0xc00051aa20}, {0x107d928, 0xc0005cb420}, {0x10709e0, 0xc00011a1e0}, 0xc0005afb48, {0x10757a8, ...}, ...}, ...}, ...)
```
Input ipv6 address: `2601:600:8a80:207e:a57d:7567:e2c9:e7b3` the Ip address is valid, but cannot map to any ipv4

### PR DESCRIPTION
**Pull request by srenatus**
There are no default prefixes in IPv6, so if an IPv6 is fed into
net.cidr_merge, we'll return a non-halt error now.

Before, we'd fail in various ways if a prefix-less IPv6 was fed into
`net.cidr_merge`. With only one, we'd return `[ "<nil>" ]`, with two,
we'd panic.

Fixes #4596.

### CODE REVIEW

File: topdown/cidr.go [hunk 1]
```diff
@@ -332,25 +332,26 @@ func evalNetCIDRMerge(networks []*net.IPNet) []*net.IPNet {
...
+		_, network, err := net.ParseCIDR(string(e))
+		return network, err
+	}
+
+	network := &net.IPNet{IP: ip}
+	switch len(ip) {
+	case net.IPv4len:
+		network.Mask = ip.DefaultMask()
+	case net.IPv6len:
+		network.Mask = net.CIDRMask(128, 128) // "/128"
```
Comment (ashutosh-narkar) [inline topdown/cidr.go:352]: If only IPv4 addresses have default masks, should we assume this or would it be better to return an error if one is not provided ?

Comment (srenatus) [reply topdown/cidr.go:352]: Yeah good point. I'll change this to return a (non-halt) error if the IPv6 input doesn't come with a prefix and can't be mapped to an IPv4 (which would have a default mask).

File: docs/content/policy-reference.md [hunk 2]
```diff
@@ -1015,7 +1015,7 @@ The table below shows examples of calling `http.send`:
...
-| <span class="opa-keep-it-together">``net.cidr_merge(cidrs_or_ips)``</span> | `output` is the smallest possible set of CIDRs obtained after merging the provided list of IP addresses and subnets in `cidrs_or_ips`  (e.g., `net.cidr_merge(["192.0.128.0/24", "192.0.129.0/24"])` generates `{"192.0.128.0/23"}`. This function merges adjacent subnets where possible, those contained within others and also removes any duplicates. Supports both IPv4 and IPv6 notations. | ``SDK-dependent`` |
+| <span class="opa-keep-it-together">``net.cidr_merge(cidrs_or_ips)``</span> | `output` is the smallest possible set of CIDRs obtained after merging the provided list of IP addresses and subnets in `cidrs_or_ips`  (e.g., `net.cidr_merge(["192.0.128.0/24", "192.0.129.0/24"])` generates `{"192.0.128.0/23"}`. This function merges adjacent subnets where possible, those contained within others and also removes any duplicates. Supports both IPv4 and IPv6 notations. For prefix-less IPv6 inputs, the prefix `/128` is assumed. | ``SDK-dependent`` |
```
Comment (ashutosh-narkar) [inline docs/content/policy-reference.md:1018]: Can we update this note as it's no longer valid.

**Review by ashutosh-narkar — DISMISSED** [review]
LGTM. Just the doc needs to be updated.

File: docs/content/policy-reference.md (hunk 2 above)
Comment (srenatus) [reply docs/content/policy-reference.md:1018]: Of course! Thanks for paying attention

**Pull request merged**